contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,180 @@
+    replicasets that create pods
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service in a loosely-coupled
+    way, so that I can concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Service Injection Policy allows pod authors to not have to explicitly
+set information for every pod. The information will be injected based off the
+policy's LabelSelector.
+
+## Proposed Changes
+
+### ServiceInjectionPolicy API object
+
+```go
+type ServiceInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec ServiceInjectionPolicySpec
+}
+
+type ServiceInjectionPolicySpec struct {
+    Service       string # Reference to the service that is being injected
+    LabelSelector *unversioned.LabelSelector
+    Env           []EnvVar
+    EnvFrom       []EnvVarFrom
+    Volumes       []Volume
+    VolumeMounts  []VolumeMount
+}
+```
+
+### AdmissionControl Plugin: ServiceInjectionPolicy
+
+The **ServiceInjectionPolicy** plugin introspects all incoming pod requests and
+injects the pod based off a `LabelSelector` with the desired attributes.
+
+#### Behavior
+
+This will modify the containers in the pod spec. The supported changes to
+`Env`, `EnvFrom`, `Volumes`, and `VolumeMounts` apply to the container spec for
+all containers in the pod with the specified matching `LabelSelector`.
+
+*Why all modify all containers in a pod?*
+
+Currently there is no concept of labels on specific containers in a pod which
+would be necessary for per-container service injections. We could add labels
+for specific containers which would allow this and be the best solution to not
+injecting all. Container labels have been discussed various times through
+multiple issues and proposals, which all congregate to this thread on the
+[kubernetes-sig-node mailing
+list](https://groups.google.com/forum/#!topic/kubernetes-sig-node/gijxbYC7HT8).
+
+Other solutions include basing the container to inject based off
+matching it's name to another field in the `ServiceInjectionPolicy` spec, but
+this would not scale well and would cause annoyance with configuration
+management.
+
+The resultant modified spec will be annotated to show that it was modified by
+the `ServiceInjectionPolicy`.
+
+#### Optional Behavior with implicit injections
+
+The above described the behavior for explicitly defining what should be
+injected into a pod.
+
+But a helper has been discussed to automatically fill in the
+`ServiceInjectionSpec` values based off annotations on a service defining the
+resources you need. This would be similar to the functionality ConfigMaps and
+Secrets.
+
+## Examples
+
+User submitted pod spec:
+
+```yaml
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: website
+spec:
+  replicas: 2
+  selector:
+    app: website
```

Comment:
My k8s newbie-ness might be showing, but if the selector is "app: website" shouldn't there be a label down below with "app: website" ?